tests/test_run_reference_vmm.py

```diff
@@ -138,6 +139,43 @@ def expect_string(vmm_process, expected_string, timeout=TEST_TIMEOUT):
             raise
 
 
+def run_cmd_inside_vm(cmd, vmm_process, prompt, timeout=5):
+    """Runs a command inside the VM process and returns the output.
+
+       If the command runs successfully, output of the command is returned or else a
+       suitable exception is raised. The `timeout` parameter is used to indicate how
+       much time to wait for the command to complete.
+
+       Note: `cmd` and `prompt` should be a `bytes` object and not `str`.
+    """
+    cmd = cmd.strip() + b'\r\n'
+
+    vmm_process.stdin.write(cmd)
+    vmm_process.stdin.flush()
+
+    then = time.time()
+    giveup_after = timeout
+    while True:
```

Comment:
I think this `while` loop can conceptually be replaced by overriding the timeout on each test that calls this function. Something like:

```python
@pytest.mark.timeout(10)
def test_reference_vmm_timeout():
  """ Times out in 10 seconds. """
  ...
  output = run_cmd_inside_vm(cmd.encode(), vmm_process, prompt.encode())
```

It'd be moving the timeout from per-command to per-test, but overall it's less verbose. WDYT?

Reply:
I am not sure whether I am getting what you are trying to say. From what I understand you suggest, rather than having a `while True....`  we can just time out the test if whatever we expect to happen does not happen. 

It's a bit difficult to  achieve, because once we set the reading to non-blocking, we've to read in some kind of loop to read what we want and we cannot 'not' set it non-blocking (for the `expect_string`) part. 

The problem with this implementation is - there is a repetition of what we are doing with `epxect_string` and `run_cmd_inside_vm`. Ideally both this should be merged into a function that gives an API that says - read everything up to timeout or read up to expected string.  That's almost like an API that `pexpect` provides - more or less. 

In general, it's a good idea to start using the `pexpect` APIs but that change is substantial (discussed in other comments.) 

Having a per test timeout is a good idea - but I believe that is provided by `pytest-timeout` which is another package (and not available with `pytest` itself.) that is not there available with the CI docker Image IIRC. 

With `pexpect` (available in CI docker image) we get nicer APIs  (running commands) and get everything that `pytest-timeout` would achieve in this particular case. 

So this `while True` kludge will go away after `pexpect`.

Reply:
Agreed - I tried a quick and dirty local refactor to see what it would look like with `pexpect` and it cleans up nicely. We can track adding `pytest-timeout` separately, it's not related to this improvement. Thanks for the explanation!